src/osd/PrimaryLogPG.cc

```diff
@@ -9976,6 +9982,280 @@ struct C_Flush : public Context {
+    return hobject_t();
+  }
+  object_t fp_oid = [&fp_algo, &chunk]() -> string {
+    switch (fp_algo) {
+      case pg_pool_t::TYPE_FINGERPRINT_SHA1:
+	return ceph::crypto::digest<ceph::crypto::SHA1>(chunk).to_str();
+      case pg_pool_t::TYPE_FINGERPRINT_SHA256:
+	return ceph::crypto::digest<ceph::crypto::SHA256>(chunk).to_str();
+      case pg_pool_t::TYPE_FINGERPRINT_SHA512:
+	return ceph::crypto::digest<ceph::crypto::SHA512>(chunk).to_str();
+      default:
+	assert(0 == "unrecognized fingerprint type");
+	return {};
+    }
+  }();    
+
+  pg_t raw_pg;
+  object_locator_t oloc(soid);
+  oloc.pool = pool.info.get_dedup_tier();
+  get_osdmap()->object_locator_to_pg(fp_oid, oloc, raw_pg);
+  hobject_t target(fp_oid, oloc.key, snapid_t(),
+		    raw_pg.ps(), raw_pg.pool(),
+		    oloc.nspace);
+  return target;
+}
+
+void PrimaryLogPG::finish_set_dedup(hobject_t oid, int r, ceph_tid_t tid, uint64_t offset)
+{
+  dout(10) << __func__ << " " << oid << " tid " << tid
+	   << " " << cpp_strerror(r) << dendl;
+  map<hobject_t,ManifestOpRef>::iterator p = manifest_ops.find(oid);
+  if (p == manifest_ops.end()) {
+    dout(10) << __func__ << " no manifest_op found" << dendl;
+    return;
+  }
+  ManifestOpRef mop = p->second;
+  mop->results[offset] = r;
+  if (r < 0) {
+    // if any failure occurs, put a mark on the results to recognize the failure
+    mop->results[0] = r;
+  }
+  if (mop->results[0] < 0) {
+    // check if the previous op returns fail
+    if (mop->num_chunks == mop->results.size()) {
+      manifest_ops.erase(oid);
+    }
+    return;
+  }
+  if (mop->num_chunks != mop->results.size()) {
+    // there are on-going works
+    return;
+  }
+  ObjectContextRef obc = get_object_context(oid, false);
+  if (!obc) {
+    if (mop->op)
+      osd->reply_op_error(mop->op, -EINVAL);
+    return;
+  }
+  ceph_assert(obc->is_blocked());
+  obc->stop_block();
+  kick_object_context_blocked(obc);
+  if (r < 0) {
+    if (mop->op)
+      osd->reply_op_error(mop->op, r);
+    return;
+  }
+
+  if (mop->chunks.size()) {
+    OpContextUPtr ctx = simple_opc_create(obc);
+    ceph_assert(ctx);
+    if (ctx->lock_manager.get_lock_type(
+	  RWState::RWWRITE,
+	  oid,
+	  obc,
+	  mop->op)) {
+      dout(20) << __func__ << " took write lock" << dendl;
+    } else if (mop->op) {
+      dout(10) << __func__ << " waiting on write lock " << mop->op << dendl;
+      close_op_ctx(ctx.release());
+      osd->reply_op_error(mop->op, -EAGAIN);
+      return;    
+    }
+
+    object_ref_delta_t refs;
+    ctx->at_version = get_next_version();
+    ctx->new_obs = obc->obs;
+    ctx->new_obs.oi.clear_flag(object_info_t::FLAG_DIRTY);
+    for (auto p : mop->chunks) {
+      struct chunk_info_t info;
+      info.offset = 0;
+      info.length = p.second.second;
+      info.oid =  p.first;
+      ctx->new_obs.oi.manifest.chunk_map[p.second.first] = info;
+      // drop all references issued before
+      refs.dec_ref(p.first);
```

Comment:
I'm not sure I understand this.  Previously, I was unsure of what would happen if we flushed an object that already had some dedup'd sections.  I didn't see any logic to clear them from the chunk_map (maybe I missed it?).  Here, we have a new problem.  You appear to be clearing references on the *new* oids (mop->chunks reflects the chunks we are now writing out, right?).  That seems wrong.

Perhaps the answer is to revert to your previous version, but clear the chunk_map before adding the mop->chunks portions back in?

Prior to writing code, please respond to this comment with an explanation of your intention so that I can better understand what you are trying to do.

Reply:
See my comment in start_dedup.

Reply:
I wrote the wrong code by mistake.

Reply:
This probably means we're missing some test coverage as well.